File: CPP/MaximumPathQualityOfAGraph.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    int maximalPathQuality(std::vector<int>& values, std::vector<std::vector<int>>& edges, int maxTime) {
        int n = (int)values.size();
        adj.assign(n, {});
        for (auto& e : edges) {
            adj[e[0]].push_back({e[1], e[2]});
            adj[e[1]].push_back({e[0], e[2]});
        }
        vals = &values;
        visited.assign(n, 0);
        best = 0;
        visited[0] = 1;
        dfs(0, maxTime, values[0]);
        return (int)best;
    }

private:
    std::vector<std::vector<std::pair<int, int>>> adj;  // node -> (to, cost)
    std::vector<int>* vals = nullptr;
    std::vector<int> visited;
    long long best = 0;

    void dfs(int node, int timeLeft, long long quality) {
        if (node == 0 && quality > best) best = quality;
        for (auto& [to, cost] : adj[node]) {
            if (cost <= timeLeft) {
                long long gain = visited[to] == 0 ? (*vals)[to] : 0;
                visited[to]++;
                dfs(to, timeLeft - cost, quality + gain);
                visited[to]--;
            }
        }
    }
};

int main() {
    Solution s;
    std::vector<int> v1 = {0, 32, 10, 43};
    std::vector<std::vector<int>> e1 = {{0, 1, 10}, {1, 2, 15}, {0, 3, 10}};
    std::vector<int> v2 = {5, 10, 15, 20};
    std::vector<std::vector<int>> e2 = {{0, 1, 10}, {1, 2, 10}, {0, 3, 10}};
    std::vector<int> v3 = {1, 2, 3, 4};
    std::vector<std::vector<int>> e3 = {{0, 1, 10}, {1, 2, 11}, {2, 3, 12}, {1, 3, 13}};
    std::vector<int> v4 = {5};
    std::vector<std::vector<int>> e4 = {};
    std::cout << s.maximalPathQuality(v1, e1, 49) << "\n";  // 75
    std::cout << s.maximalPathQuality(v2, e2, 30) << "\n";  // 25
    std::cout << s.maximalPathQuality(v3, e3, 50) << "\n";  // 7
    std::cout << s.maximalPathQuality(v4, e4, 100) << "\n"; // 5
    return 0;
}

File: Rust/src/bin/maximum_path_quality_of_a_graph.rs
fn dfs(
    node: usize,
    time_left: i32,
    quality: i64,
    adj: &Vec<Vec<(usize, i32)>>,
    values: &Vec<i64>,
    visited: &mut Vec<i32>,
    best: &mut i64,
) {
    if node == 0 && quality > *best {
        *best = quality;
    }
    for &(to, cost) in &adj[node] {
        if cost <= time_left {
            let gain = if visited[to] == 0 { values[to] } else { 0 };
            visited[to] += 1;
            dfs(to, time_left - cost, quality + gain, adj, values, visited, best);
            visited[to] -= 1;
        }
    }
}

fn maximal_path_quality(values: Vec<i32>, edges: Vec<Vec<i32>>, max_time: i32) -> i32 {
    let n = values.len();
    let vals: Vec<i64> = values.iter().map(|&x| x as i64).collect();
    let mut adj: Vec<Vec<(usize, i32)>> = vec![Vec::new(); n];
    for e in &edges {
        let (u, v, t) = (e[0] as usize, e[1] as usize, e[2]);
        adj[u].push((v, t));
        adj[v].push((u, t));
    }
    let mut visited = vec![0i32; n];
    let mut best = 0i64;
    visited[0] = 1;
    dfs(0, max_time, vals[0], &adj, &vals, &mut visited, &mut best);
    best as i32
}

fn main() {
    println!("{}", maximal_path_quality(vec![0, 32, 10, 43], vec![vec![0, 1, 10], vec![1, 2, 15], vec![0, 3, 10]], 49)); // 75
    println!("{}", maximal_path_quality(vec![5, 10, 15, 20], vec![vec![0, 1, 10], vec![1, 2, 10], vec![0, 3, 10]], 30)); // 25
    println!("{}", maximal_path_quality(vec![1, 2, 3, 4], vec![vec![0, 1, 10], vec![1, 2, 11], vec![2, 3, 12], vec![1, 3, 13]], 50)); // 7
    println!("{}", maximal_path_quality(vec![5], vec![], 100)); // 5
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn examples() {
        assert_eq!(maximal_path_quality(vec![0, 32, 10, 43], vec![vec![0, 1, 10], vec![1, 2, 15], vec![0, 3, 10]], 49), 75);
        assert_eq!(maximal_path_quality(vec![5, 10, 15, 20], vec![vec![0, 1, 10], vec![1, 2, 10], vec![0, 3, 10]], 30), 25);
        assert_eq!(maximal_path_quality(vec![1, 2, 3, 4], vec![vec![0, 1, 10], vec![1, 2, 11], vec![2, 3, 12], vec![1, 3, 13]], 50), 7);
        assert_eq!(maximal_path_quality(vec![5], vec![], 100), 5);
    }
}
